README.md
<p align="center">
<img src="https://raw.githubusercontent.com/estripling/sparkit/main/docs/source/_static/logo.png" width="480" alt="The sparkit logo.">
</p>

<p align="center">
<a href="https://pypi.org/project/sparkit"><img alt="pypi" src="https://img.shields.io/pypi/v/sparkit"></a>
<a href="https://readthedocs.org/projects/sparkit/?badge=latest"><img alt="docs" src="https://readthedocs.org/projects/sparkit/badge/?version=latest"></a>
<a href="https://github.com/estripling/sparkit/actions/workflows/ci.yml"><img alt="ci status" src="https://github.com/estripling/sparkit/actions/workflows/ci.yml/badge.svg?branch=main"></a>
<a href="https://codecov.io/gh/estripling/sparkit"><img alt="coverage" src="https://codecov.io/github/estripling/sparkit/coverage.svg?branch=main"></a>
<a href="https://github.com/estripling/sparkit/blob/main/LICENSE"><img alt="license" src="https://img.shields.io/pypi/l/sparkit"></a>
</p>

## About

A package for PySpark utility functions:

- [Documentation](https://sparkit.readthedocs.io/en/stable/index.html)
- [Example usage](https://sparkit.readthedocs.io/en/stable/example.html)
- [API Reference](https://sparkit.readthedocs.io/en/stable/autoapi/sparkit/index.html)

## Installation

`sparkit` is available on [PyPI](https://pypi.org/project/sparkit/) for Python 3.8+ and Spark 3 (Java 11):

```console
pip install sparkit
```

## Examples

[`union`](https://sparkit.readthedocs.io/en/stable/autoapi/sparkit/index.html#sparkit.union) multiple data frames by name (pass single and / or an iterable of data frames):

```python
>>> import sparkit
>>> from pyspark.sql import Row, SparkSession
>>> spark = SparkSession.builder.getOrCreate()
>>> df1 = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
>>> df2 = spark.createDataFrame([Row(x=5, y=6), Row(x=7, y=8)])
>>> df3 = spark.createDataFrame([Row(x=0, y=1), Row(x=2, y=3)])
>>> df4 = spark.createDataFrame([Row(x=5, y=3), Row(x=9, y=6)])
>>> sparkit.union(df1, [df2, df3], df4).show()
+---+---+
|  x|  y|
+---+---+
|  1|  2|
|  3|  4|
|  5|  6|
|  7|  8|
|  0|  1|
|  2|  3|
|  5|  3|
|  9|  6|
+---+---+
```

## Contributing to sparkit

Your contribution is greatly appreciated!
See the following links to help you get started:

- [Contributing Guide](https://sparkit.readthedocs.io/en/latest/contributing.html)
- [Developer Guide](https://sparkit.readthedocs.io/en/latest/developers.html)
- [Contributor Code of Conduct](https://sparkit.readthedocs.io/en/latest/conduct.html)

## License

`sparkit` was created by sparkit Developers.
It is licensed under the terms of the BSD 3-Clause license.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.1 (2023-09-26)


## v1.1.0 (2023-04-30)
### Feature
* Add pandastools ([#15](https://github.com/estripling/sparkit/issues/15)) ([`0c0aff6`](https://github.com/estripling/sparkit/commit/0c0aff662ef9ce3c379d0ee249f9b7ed64bf385e))

### Fix
* Readthedocs failed built ([`917422c`](https://github.com/estripling/sparkit/commit/917422cc99bb4fea3da537c5f70c71a5ea219eb7))

## v1.0.0 (2023-04-23)
### Feature
* Add time related functions ([#13](https://github.com/estripling/sparkit/issues/13)) ([`5e710e2`](https://github.com/estripling/sparkit/commit/5e710e2ca082370fb262db736034f6aa8756992c))

### Breaking
* new function signatures and renamed function ([`937891b`](https://github.com/estripling/sparkit/commit/937891bb65bd603568c09d9f16c35d042a53b2f0))

### Documentation
* Refactor function signatures ([#14](https://github.com/estripling/sparkit/issues/14)) ([`937891b`](https://github.com/estripling/sparkit/commit/937891bb65bd603568c09d9f16c35d042a53b2f0))

## v0.3.0 (2023-03-20)
### Feature
* Add with_index function ([#12](https://github.com/estripling/sparkit/issues/12)) ([`0f58112`](https://github.com/estripling/sparkit/commit/0f5811294568d12b7793c2452ae5ade411338b7b))

## v0.2.0 (2023-03-13)
### Feature
* Add peek function ([#9](https://github.com/estripling/sparkit/issues/9)) ([`d3e180f`](https://github.com/estripling/sparkit/commit/d3e180fa535527f5984085a81d1f04088313d0b2))
* Add freq function ([#8](https://github.com/estripling/sparkit/issues/8)) ([`d7f728c`](https://github.com/estripling/sparkit/commit/d7f728cf9ffca2f9e38e5a951147cd48e0e027ab))
* Add prefix suffix functions ([#7](https://github.com/estripling/sparkit/issues/7)) ([`445cff0`](https://github.com/estripling/sparkit/commit/445cff0e26d2bbb11c05c8a21ba8fb4dfa2447e9))
* Add count_nulls function ([#6](https://github.com/estripling/sparkit/issues/6)) ([`64a7077`](https://github.com/estripling/sparkit/commit/64a70774c3cc532bd92617b7734218b99924ff69))
* Add join function ([#5](https://github.com/estripling/sparkit/issues/5)) ([`b8aae15`](https://github.com/estripling/sparkit/commit/b8aae150f8828c17103ace095b09a4b32c736418))
* Add union function ([#4](https://github.com/estripling/sparkit/issues/4)) ([`5dbc6dd`](https://github.com/estripling/sparkit/commit/5dbc6dde68301c147a47eae41afcdafbbaddcc17))

## v0.1.0 (2023-03-11)
### Feature
* Add dataframe validation functions ([#3](https://github.com/estripling/sparkit/issues/3)) ([`9ce8640`](https://github.com/estripling/sparkit/commit/9ce86404b366d69e11547fdbead63d19f5f28f25))

## v0.0.2 (2023-02-12)
### Fix
* Move logos to right directory ([`362a889`](https://github.com/estripling/sparkit/commit/362a889dd4628a91339593a8d32fdc0ed96cac2a))

## v0.0.1 (2023-02-12)
### Documentation
* Add logo ([`04681f9`](https://github.com/estripling/sparkit/commit/04681f9d4d70c86e27011ade9ac9393f416beb0f))


tests/conftest.py
import pytest
from pyspark.sql import DataFrame, SparkSession


def assert_dataframe_equal(lft_df: DataFrame, rgt_df: DataFrame):
    """Assert that the left and right data frames are equal."""
    lft_schema = lft_df.schema.simpleString()
    rgt_schema = rgt_df.schema.simpleString()
    assert lft_schema == rgt_schema, "schema mismatch"

    assert lft_df.count() == rgt_df.count(), "row count mismatch"

    lft_rows = lft_df.subtract(rgt_df)
    rgt_rows = rgt_df.subtract(lft_df)
    assert (lft_rows.count() == 0) and (rgt_rows.count() == 0), "row mismatch"


@pytest.fixture(scope="session")
def spark():
    spark = (
        SparkSession.builder.master("local[1]")
        .appName("spark-session-for-tests")
        .config("spark.driver.memory", "2g")
        .config("spark.executor.cores", 1)
        .config("spark.executor.instances", 1)
        .config("spark.sql.shuffle.partitions", 1)
        .config("spark.ui.enabled", False)
        .config("spark.ui.showConsoleProgress", False)
        .getOrCreate()
    )
    yield spark
    spark.stop()


tests/test_sparkit.py
from importlib import metadata

from pyspark.sql import SparkSession

import sparkit


def test_version():
    assert sparkit.__version__ == metadata.version("sparkit")


def test_spark_session(spark):
    assert isinstance(spark, SparkSession)


tests/test_validation.py
import pytest
from pyspark.sql import Row

import sparkit


def test_check_row_count_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1), Row(x=3)])
    rgt_df__different = spark.createDataFrame([Row(x=1)])

    assert sparkit.check_row_count_equal(lft_df, rgt_df__equal) is None

    with pytest.raises(sparkit.exception.RowCountMismatchError):
        sparkit.check_row_count_equal(lft_df, rgt_df__different)


def test_check_row_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
    rgt_df__different = spark.createDataFrame([Row(x=1, y=7), Row(x=3, y=9)])

    assert sparkit.check_row_equal(lft_df, rgt_df__equal) is None

    with pytest.raises(sparkit.exception.RowMismatchError):
        sparkit.check_row_equal(lft_df, rgt_df__different)


def test_check_schema_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
    rgt_df__different_type = spark.createDataFrame([Row(x=1, y="2"), Row(x=3, y="4")])
    rgt_df__different_size = spark.createDataFrame([Row(x=1), Row(x=3)])

    assert sparkit.check_schema_equal(lft_df, rgt_df__equal) is None

    with pytest.raises(sparkit.exception.SchemaMismatchError):
        sparkit.check_schema_equal(lft_df, rgt_df__different_type)

    with pytest.raises(sparkit.exception.SchemaMismatchError):
        sparkit.check_schema_equal(lft_df, rgt_df__different_size)


def test_is_row_count_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1), Row(x=3)])
    rgt_df__different = spark.createDataFrame([Row(x=1)])

    assert sparkit.is_row_count_equal(lft_df, rgt_df__equal)
    assert not sparkit.is_row_count_equal(lft_df, rgt_df__different)


def test_is_row_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
    rgt_df__different = spark.createDataFrame([Row(x=1, y=7), Row(x=3, y=9)])

    assert sparkit.is_row_equal(lft_df, rgt_df__equal)
    assert not sparkit.is_row_equal(lft_df, rgt_df__different)


def test_is_schema_equal(spark):
    lft_df = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])

    rgt_df__equal = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
    rgt_df__different_type = spark.createDataFrame([Row(x=1, y="2"), Row(x=3, y="4")])
    rgt_df__different_size = spark.createDataFrame([Row(x=1), Row(x=3)])

    assert sparkit.is_schema_equal(lft_df, rgt_df__equal)
    assert not sparkit.is_schema_equal(lft_df, rgt_df__different_type)
    assert not sparkit.is_schema_equal(lft_df, rgt_df__different_size)


src/sparkit/exception.py
import bumbag


class SparkitError(Exception):
    """A base class for sparkit exceptions."""

    pass


class RowCountMismatchError(SparkitError):
    """Exception raised for mismatching row counts.

    Parameters
    ----------
    lft_row_count : int
        Left row count.
    rgt_row_count : int
        Right row count.
    """

    def __init__(self, lft_row_count, rgt_row_count):
        self.lft_row_count = lft_row_count
        self.rgt_row_count = rgt_row_count
        self.difference = lft_row_count - rgt_row_count
        self.ratio = lft_row_count / rgt_row_count
        self.message = (
            f"{lft_row_count=:,}\n"
            + f"{rgt_row_count=:,}\n"
            + f"{lft_row_count - rgt_row_count=:,}\n"
            + f"{lft_row_count / rgt_row_count=:g}"
        )
        super().__init__(self.message)


class SchemaMismatchError(SparkitError):
    """Exception raised for mismatching schemas.

    Parameters
    ----------
    lft_schema : str
        Left schema.
    rgt_schema : str
        Right schema.
    """

    def __init__(self, lft_schema, rgt_schema):
        self.lft_schema = lft_schema
        self.rgt_schema = rgt_schema
        msg = bumbag.highlight_string_differences(lft_schema, rgt_schema)
        num_character_differences = sum(c == "|" for c in msg.splitlines()[1])
        self.message = f"{num_character_differences=}\n{msg}"
        super().__init__(self.message)


src/sparkit/transformation.py
import functools

import bumbag
import pyspark.sql.functions as F
import pyspark.sql.types as T
import toolz
from IPython import get_ipython
from IPython.display import HTML, display
from pyspark.sql import DataFrame, Window

__all__ = (
    "add_prefix",
    "add_suffix",
    "count_nulls",
    "daterange",
    "freq",
    "join",
    "peek",
    "union",
    "with_consecutive_integers",
    "with_endofweek_date",
    "with_startofweek_date",
    "with_weekday_name",
)


@toolz.curry
def add_prefix(prefix, dataframe, /, *, subset=None):
    """Add prefix to column names.

    Parameters
    ----------
    prefix : str
        Specify the prefix string.
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    subset : Iterable of str, default=None
        Specify a column selection. If None, all columns are selected.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with changed column names.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame([Row(x=1, y=2)])
    >>> sparkit.add_prefix("prefix_", df).show()
    +--------+--------+
    |prefix_x|prefix_y|
    +--------+--------+
    |       1|       2|
    +--------+--------+
    <BLANKLINE>
    """
    columns = subset or dataframe.columns
    for column in columns:
        dataframe = dataframe.withColumnRenamed(column, f"{prefix}{column}")
    return dataframe


@toolz.curry
def add_suffix(suffix, dataframe, /, *, subset=None):
    """Add suffix to column names.

    Parameters
    ----------
    suffix : str
        Specify the suffix string.
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    subset : Iterable of str, default=None
        Specify a column selection. If None, all columns are selected.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with changed column names.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame([Row(x=1, y=2)])
    >>> sparkit.add_suffix("_suffix", df).show()
    +--------+--------+
    |x_suffix|y_suffix|
    +--------+--------+
    |       1|       2|
    +--------+--------+
    <BLANKLINE>
    """
    columns = subset or dataframe.columns
    for column in columns:
        dataframe = dataframe.withColumnRenamed(column, f"{column}{suffix}")
    return dataframe


@toolz.curry
def count_nulls(dataframe, /, *, subset=None):
    """Count null values in data frame.

    Parameters
    ----------
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    subset : Iterable of str, default=None
        Specify a column selection. If None, all columns are selected.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with null value counts per column.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(x=1, y=2, z=None),
    ...         Row(x=4, y=None, z=6),
    ...         Row(x=10, y=None, z=None),
    ...     ]
    ... )
    >>> sparkit.count_nulls(df).show()
    +---+---+---+
    |  x|  y|  z|
    +---+---+---+
    |  0|  2|  2|
    +---+---+---+
    <BLANKLINE>
    """
    columns = subset or dataframe.columns
    return dataframe.agg(
        *[F.sum(F.isnull(c).cast(T.LongType())).alias(c) for c in columns]
    )


@toolz.curry
def daterange(id_column_name, new_column_name, dataframe, /, *, min_date, max_date):
    """Generate a date range for each distinct ID value.

    Parameters
    ----------
    id_column_name : str
        Specify the name of the ID column.
    new_column_name : str
        Specify the name of the new column to be added to the data frame.
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    min_date : str or datetime.date
        Specify the inclusive lower endpoint of the date range.
    max_date : str or datetime.date
        Specify the inclusive upper endpoint of the date range.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame in long format with consecutive dates between
        `min_date` and `max_date` for each distinct ID value.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(id=1),
    ...         Row(id=1),
    ...         Row(id=3),
    ...         Row(id=2),
    ...         Row(id=2),
    ...         Row(id=3),
    ...     ]
    ... )
    >>> (
    ...     sparkit.daterange(
    ...         "id",
    ...         "day",
    ...         df,
    ...         min_date="2023-05-01",
    ...         max_date="2023-05-03",
    ...     )
    ...     .orderBy("id", "day")
    ...     .show()
    ... )
    +---+----------+
    | id|       day|
    +---+----------+
    |  1|2023-05-01|
    |  1|2023-05-02|
    |  1|2023-05-03|
    |  2|2023-05-01|
    |  2|2023-05-02|
    |  2|2023-05-03|
    |  3|2023-05-01|
    |  3|2023-05-02|
    |  3|2023-05-03|
    +---+----------+
    <BLANKLINE>
    """
    return (
        dataframe.select(id_column_name)
        .distinct()
        .withColumn("min_date", F.to_date(F.lit(min_date), "yyyy-MM-dd"))
        .withColumn("max_date", F.to_date(F.lit(max_date), "yyyy-MM-dd"))
        .select(
            id_column_name,
            F.expr("sequence(min_date, max_date, interval 1 day)").alias(
                new_column_name
            ),
        )
        .withColumn(new_column_name, F.explode(new_column_name))
    )


@toolz.curry
def freq(columns, dataframe, /):
    """Compute value frequencies.

    Given a selection of columns, calculate for each distinct value:
     - the frequency (``frq``),
     - the cumulative frequency (``cml_frq``),
     - the relative frequency (``rel_frq``), and
     - the cumulative relative frequency (``rel_cml_frq``).

    Parameters
    ----------
    columns : Iterable of str or pyspark.sql.Column
        Specify the columns for which to compute the value frequency.
    dataframe : pyspark.sql.DataFrame
        Input data frame.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with value frequencies of specified columns.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(x="a"),
    ...         Row(x="c"),
    ...         Row(x="b"),
    ...         Row(x="g"),
    ...         Row(x="h"),
    ...         Row(x="a"),
    ...         Row(x="g"),
    ...         Row(x="a"),
    ...     ]
    ... )
    >>> sparkit.freq(["x"], df).show()
    +---+---+-------+-------+-----------+
    |  x|frq|cml_frq|rel_frq|rel_cml_frq|
    +---+---+-------+-------+-----------+
    |  a|  3|      3|  0.375|      0.375|
    |  g|  2|      5|   0.25|      0.625|
    |  b|  1|      6|  0.125|       0.75|
    |  c|  1|      7|  0.125|      0.875|
    |  h|  1|      8|  0.125|        1.0|
    +---+---+-------+-------+-----------+
    <BLANKLINE>
    """
    # Use F.lit(1) for an ungrouped specification
    win_sorted = Window.partitionBy(F.lit(1)).orderBy(F.desc("frq"), *columns)
    win_unsorted = Window.partitionBy(F.lit(1))
    return (
        dataframe.groupby(columns)
        .count()
        .withColumnRenamed("count", "frq")
        .withColumn("cml_frq", F.sum("frq").over(win_sorted))
        .withColumn("rel_frq", F.col("frq") / F.sum("frq").over(win_unsorted))
        .withColumn("rel_cml_frq", F.sum("rel_frq").over(win_sorted))
        .orderBy("cml_frq")
    )


def join(*dataframes, on, how="inner"):
    """Join multiple spark data frames on common key.

    Parameters
    ----------
    dataframes : Iterable of pyspark.sql.DataFrame
        Data frames to join.
    on : str or iterable of str
        Key(s) to join on.
    how : str, default="inner"
        Valid specification to join spark data frames.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame being the join of supplied data frames.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df1 = spark.createDataFrame([Row(id=1, x="a"), Row(id=2, x="b")])
    >>> df2 = spark.createDataFrame([Row(id=1, y="c"), Row(id=2, y="d")])
    >>> df3 = spark.createDataFrame([Row(id=1, z="e"), Row(id=2, z="f")])
    >>> sparkit.join(df1, df2, df3, on="id").show()
    +---+---+---+---+
    | id|  x|  y|  z|
    +---+---+---+---+
    |  1|  a|  c|  e|
    |  2|  b|  d|  f|
    +---+---+---+---+
    <BLANKLINE>
    """
    join = functools.partial(DataFrame.join, on=on, how=how)
    return functools.reduce(join, bumbag.flatten(dataframes))


@toolz.curry
def peek(dataframe, /, *, n=6, cache=False, schema=False, index=False):
    """Have a quick look at the data frame and return it.

    This function is handy when chaining data frame transformations.

    Parameters
    ----------
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    n : int, default=6
        Specify the number of rows to show. If `n <= 0`, no rows are shown.
    cache : bool, default=False
        Specify if data frame should be cached.
    schema : bool, default=False
        Specify if schema should be printed.
    index : bool, default=False
        Specify if a row index should be shown.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        The input data frame.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(x=1, y="a"),
    ...         Row(x=3, y=None),
    ...         Row(x=None, y="c"),
    ...     ]
    ... )
    >>> df.show()
    +----+----+
    |   x|   y|
    +----+----+
    |   1|   a|
    |   3|null|
    |null|   c|
    +----+----+
    <BLANKLINE>
    >>> filtered_df = (
    ...     df.transform(sparkit.peek)
    ...     .where("x IS NOT NULL")
    ...     .transform(sparkit.peek)
    ... )
    shape = (3, 2)
       x    y
     1.0    a
     3.0 None
    None    c
    shape = (2, 2)
     x    y
     1    a
     3 None
    """
    df = dataframe if dataframe.is_cached else dataframe.cache() if cache else dataframe

    if schema:
        df.printSchema()

    num_rows = bumbag.numberformat(df.count())
    num_cols = bumbag.numberformat(len(df.columns))
    print(f"shape = ({num_rows}, {num_cols})")

    if n > 0:
        pandas_df = df.limit(n).toPandas()
        pandas_df.index += 1

        is_inside_notebook = get_ipython() is not None

        df_repr = (
            pandas_df.to_html(index=index, na_rep="None", col_space="20px")
            if is_inside_notebook
            else pandas_df.to_string(index=index, na_rep="None")
        )

        display(HTML(df_repr)) if is_inside_notebook else print(df_repr)

    return df


def union(*dataframes):
    """Union multiple spark data frames by name.

    Parameters
    ----------
    dataframes : Iterable of pyspark.sql.DataFrame
        Data frames to union by name.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame containing the union of rows of supplied data frames.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df1 = spark.createDataFrame([Row(x=1, y=2), Row(x=3, y=4)])
    >>> df2 = spark.createDataFrame([Row(x=5, y=6), Row(x=7, y=8)])
    >>> df3 = spark.createDataFrame([Row(x=0, y=1), Row(x=2, y=3)])
    >>> sparkit.union(df1, df2, df3).show()
    +---+---+
    |  x|  y|
    +---+---+
    |  1|  2|
    |  3|  4|
    |  5|  6|
    |  7|  8|
    |  0|  1|
    |  2|  3|
    +---+---+
    <BLANKLINE>
    """
    return functools.reduce(DataFrame.unionByName, bumbag.flatten(dataframes))


@toolz.curry
def with_consecutive_integers(new_column_name, dataframe, /):
    """Add column with consecutive positive integers.

    Parameters
    ----------
    new_column_name : str
        Specify the name of the new column to be added to the data frame.
    dataframe : pyspark.sql.DataFrame
        Input data frame.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with a column of consecutive positive integers.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame([Row(x="a"), Row(x="b"), Row(x="c"), Row(x="d")])
    >>> sparkit.with_consecutive_integers("idx", df).show()
    +---+---+
    |  x|idx|
    +---+---+
    |  a|  1|
    |  b|  2|
    |  c|  3|
    |  d|  4|
    +---+---+
    <BLANKLINE>
    """
    win = Window.partitionBy(F.lit(1)).orderBy(F.monotonically_increasing_id())
    return dataframe.withColumn(new_column_name, F.row_number().over(win))


@toolz.curry
def with_endofweek_date(
    date_column_name,
    new_column_name,
    dataframe,
    /,
    *,
    last_weekday_name="Sun",
):
    """Add column with the end of the week date.

    Parameters
    ----------
    date_column_name : str
        Specify the name of the date column.
    new_column_name : str
        Specify the name of the new column to be added to the data frame.
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    last_weekday_name : str, default="Sun"
        Specify the name of the last weekday.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with the end-of-week date column.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(day="2023-05-01"),
    ...         Row(day=None),
    ...         Row(day="2023-05-03"),
    ...         Row(day="2023-05-08"),
    ...         Row(day="2023-05-21"),
    ...     ],
    ... )
    >>> sparkit.with_endofweek_date("day", "endofweek", df).show()
    +----------+----------+
    |       day| endofweek|
    +----------+----------+
    |2023-05-01|2023-05-07|
    |      null|      null|
    |2023-05-03|2023-05-07|
    |2023-05-08|2023-05-14|
    |2023-05-21|2023-05-21|
    +----------+----------+
    <BLANKLINE>

    >>> sparkit.with_endofweek_date(
    ...     "day", "endofweek", df, last_weekday_name="Sat"
    ... ).show()
    +----------+----------+
    |       day| endofweek|
    +----------+----------+
    |2023-05-01|2023-05-06|
    |      null|      null|
    |2023-05-03|2023-05-06|
    |2023-05-08|2023-05-13|
    |2023-05-21|2023-05-27|
    +----------+----------+
    <BLANKLINE>
    """
    tmp_column = "weekday"
    return (
        dataframe.transform(with_weekday_name(date_column_name, tmp_column))
        .withColumn(
            new_column_name,
            F.when(F.col(tmp_column).isNull(), None)
            .when(F.col(tmp_column) == last_weekday_name, F.col(date_column_name))
            .otherwise(F.next_day(F.col(date_column_name), last_weekday_name)),
        )
        .drop(tmp_column)
    )


@toolz.curry
def with_startofweek_date(
    date_column_name,
    new_column_name,
    dataframe,
    /,
    *,
    last_weekday_name="Sun",
):
    """Add column with the start of the week date.

    Parameters
    ----------
    date_column_name : str
        Specify the name of the date column.
    new_column_name : str
        Specify the name of the new column to be added to the data frame.
    dataframe : pyspark.sql.DataFrame
        Input data frame.
    last_weekday_name : str, default="Sun"
        Specify the name of the last weekday.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with the start-of-week date column.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [
    ...         Row(day="2023-05-01"),
    ...         Row(day=None),
    ...         Row(day="2023-05-03"),
    ...         Row(day="2023-05-08"),
    ...         Row(day="2023-05-21"),
    ...     ],
    ... )
    >>> sparkit.with_startofweek_date("day", "startofweek", df).show()
    +----------+-----------+
    |       day|startofweek|
    +----------+-----------+
    |2023-05-01| 2023-05-01|
    |      null|       null|
    |2023-05-03| 2023-05-01|
    |2023-05-08| 2023-05-08|
    |2023-05-21| 2023-05-15|
    +----------+-----------+
    <BLANKLINE>

    >>> sparkit.with_startofweek_date(
    ...     "day", "startofweek", df, last_weekday_name="Sat"
    ... ).show()
    +----------+-----------+
    |       day|startofweek|
    +----------+-----------+
    |2023-05-01| 2023-04-30|
    |      null|       null|
    |2023-05-03| 2023-04-30|
    |2023-05-08| 2023-05-07|
    |2023-05-21| 2023-05-21|
    +----------+-----------+
    <BLANKLINE>
    """
    tmp_column = "endofweek"
    with_endofweek = with_endofweek_date(
        date_column_name,
        tmp_column,
        last_weekday_name=last_weekday_name,
    )
    return (
        dataframe.transform(with_endofweek)
        .withColumn(new_column_name, F.date_sub(tmp_column, 6))
        .drop(tmp_column)
    )


@toolz.curry
def with_weekday_name(date_column_name, new_column_name, dataframe, /):
    """Add column with the name of the weekday.

    Parameters
    ----------
    date_column_name : str
        Specify the name of the date column.
    new_column_name : str
        Specify the name of the new column to be added to the data frame.
    dataframe : pyspark.sql.DataFrame
        Input data frame.

    Notes
    -----
    Function is curried.

    Returns
    -------
    pyspark.sql.DataFrame
        A new data frame with the weekday column.

    Examples
    --------
    >>> import sparkit
    >>> from pyspark.sql import Row, SparkSession
    >>> spark = SparkSession.builder.getOrCreate()
    >>> df = spark.createDataFrame(
    ...     [Row(day="2023-05-01"), Row(day=None), Row(day="2023-05-03")]
    ... )
    >>> sparkit.with_weekday_name("day", "weekday", df).show()
    +----------+-------+
    |       day|weekday|
    +----------+-------+
    |2023-05-01|    Mon|
    |      null|   null|
    |2023-05-03|    Wed|
    +----------+-------+
    <BLANKLINE>
    """

    def determine_weekday(date_column):
        weekday_int = F.dayofweek(date_column)
        return (
            F.when(weekday_int == 1, "Sun")
            .when(weekday_int == 2, "Mon")
            .when(weekday_int == 3, "Tue")
            .when(weekday_int == 4, "Wed")
            .when(weekday_int == 5, "Thu")
            .when(weekday_int == 6, "Fri")
            .when(weekday_int == 7, "Sat")
            .otherwise(None)
        )

    return dataframe.withColumn(new_column_name, determine_weekday(date_column_name))
